### PR TITLE
Psalm Integration

### LINKED ISSUE #1: Psalm integration
### Feature Request

|    Q        |   A
|------------ | ------
| QA    | yes

#### Summary

As decided during the [Technical-Steering-Committee Meeting on August 3rd, 2020](https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-08-03-TSC-Minutes.md#vote-on-static-analysis-tooling), Laminas wants to implement [vimeo/psalm](https://github.com/vimeo/psalm) in all packages.

Implementing psalm is quite easy.

#### Required
- [x] Create a `psalm.xml` in the project root
- [x] Copy and paste the contents from this [psalm.xml.dist](https://github.com/mezzio/repo-template/blob/6124090610ef1c6d6b0c80a063cafacc3ec68f38/psalm.xml.dist)
- [x] Run `$ composer require --dev vimeo/psalm`
- [x] Run `$ vendor/bin/psalm --set-baseline=psalm-baseline.xml`
- [x] Add a composer script `static-analysis` with the command `psalm --shepherd --stats`
- [x] Add a new line to `script:` in `.travis.yml`: `- if [[ $TEST_COVERAGE == 'true' ]]; then composer static-analysis ; fi`
- [x] Remove phpstan from the project (`phpstan.neon.dist`, `.travis.yml` entry, `composer.json` `require-dev` and `scripts`)

##### Optional
- [x] Fix as many psalm errors as possible.

### PR DESCRIPTION
**Pull request by diwms**
Hi there 👋🏻 
I'm trying to contribute and close issue #1 

I've no idea what I need to write in the issue template since this is an attempt to close the existing issue 😄 

Please, let me know what do you think guys 👍🏻 

Happy Hacktoberfest! 🥇

### CODE REVIEW

File: composer.json
Comment (boesing) [inline composer.json]: There is no need to add `--config` to the script as it will automatically use the projects `.dist` if exists.

File: src/BasicAccessFactory.php
Comment (boesing) [inline src/BasicAccessFactory.php:21]: You might want to require `webmozart/assert` and use `Assert::nullOrIsInstanceOfAny()` instead of inline annotation 👍

File: src/BasicAccessFactory.php
Comment (boesing) [inline src/BasicAccessFactory.php:33]: You might want to require webmozart/assert and use Assert::nullOrString() instead of inline annotation 👍

File: src/BasicAccessFactory.php
Comment (boesing) [inline src/BasicAccessFactory.php:43]: You might want to require webmozart/assert and use Assert::isInstanceOf() instead of inline annotation 👍

File: test/BasicAccessFactoryTest.php
Comment (boesing) [inline test/BasicAccessFactoryTest.php]: We should not drop the mocked type to avoid confusion.
What psalm error made you change these annotations?

File: test/BasicAccessTest.php
Comment (boesing) [inline test/BasicAccessTest.php]: There should be no need to annotate this inline as when you re-add property annotations properly, psalm should already know whats being revealed.

**Review by boesing — CHANGES_REQUESTED** [review]

File: composer.json
Comment (diwms) [reply composer.json]: I see. I've renamed `.psalm.xml.dist` -> `psalm.xml.dist` (removed leading dot).
Will push it soon

File: src/BasicAccessFactory.php
Comment (diwms) [reply src/BasicAccessFactory.php:33]: It will be BC change, isn't it?

`webmozart/assert` will throw `\InvalidArgumentException`
In current code, we throw own package exception `Mezzio\Authentication\Exception\InvalidConfigException`

File: src/BasicAccessFactory.php
Comment (diwms) [reply src/BasicAccessFactory.php:43]: It will be BC change, isn't it?

`webmozart/assert` will throw `\InvalidArgumentException`
In current code, we throw own package exception `Mezzio\Authentication\Exception\InvalidConfigException`

File: test/BasicAccessFactoryTest.php
Comment (diwms) [reply test/BasicAccessFactoryTest.php]: In the case of `ContainerInterface|ObjectProphecy` annotation Psalm will throw this errorr:

```
ERROR: PossiblyUndefinedMethod - test\BasicAccessFactoryTest.php:66:49 - Method Psr\Container\ContainerInterface::reveal does not exist (see https://psalm.dev/108)
        $containerInterface = $this->container->reveal();
```
The reason that there is not a common interface for these classes. and there is no `reveal` method in `ContainerInterface` as well as `willReturn` etc.

File: src/BasicAccessFactory.php
Comment (diwms) [reply src/BasicAccessFactory.php:21]: It will be BC change, isn't it?

webmozart/assert will throw \InvalidArgumentException
In current code, we throw own package exception Mezzio\Authentication\Exception\InvalidConfigException

File: test/BasicAccessTest.php
Comment (diwms) [reply test/BasicAccessTest.php]: Would be very nice if you could suggest here something.
I've no clue why psalm cannot understand mixed annotations properly except one reason:

Too old PHPUnit version. Fo correct understanding mocks psalm using [an additional plugin](https://github.com/psalm/psalm-plugin-phpunit) and in stubs directory, I see there is Prophecy.phpstub

But it looks like Prophecy.phpstub isn't suitable for PHPUnit 7.5

**Comment by diwms** [comment]
@boesing I believe all these issues with annotations can be fixed by upgrading phpunit to 9 version.
Psalm plugin then should better understand what is going on and not complaint about types and methods on prophecy mocks.

Can I also include this upgrade in this PR and close #2 as well ?
Or it's better to start a new branch with PHP 8 support and then come back to this PR?

**Comment by geerteltink** [comment]
> Can I also include this upgrade in this PR and close #2 as well ?
> Or it's better to start a new branch with PHP 8 support and then come back to this PR?

Please create a new PR for #2. In there you change the PHP constraints and upgrade phpunit.

File: src/BasicAccessFactory.php
Comment (Ocramius) [reply src/BasicAccessFactory.php:21]: At this stage, it isn't a BC break: we are silently assuming the type, but that's unsafe already.

File: src/BasicAccessFactory.php
Comment (Ocramius) [reply src/BasicAccessFactory.php:33]: Similar to above feedback: right now we **assume** `null|string`, but so many things can go wrong if it is not a `string` - instead, an exception is already a good improvement

File: src/BasicAccessFactory.php
Comment (Ocramius) [reply src/BasicAccessFactory.php:43]: I think the assertion is not needed here: there's an explicit `if ()` block after this, so the `/** @var` block can also be removed in this case.

File: test/BasicAccessFactoryTest.php
Comment (Ocramius) [reply test/BasicAccessFactoryTest.php]: @boesing this is an inherent problem of prophecy itself: the best approach here would be to replace prophecy with internal phpunit mocks, which come with proper typing too.

Up to @diwms if they want to do that or call this "done" here, because it is a bit more work.

File: test/BasicAccessTest.php
Comment (Ocramius) [reply test/BasicAccessTest.php]: The type inference works with internal phpunit mocks, but not  with prophecy, so the same feedback as above stands: IMO we should just phase out prophecy.

**Comment by Ocramius** [comment]
@diwms I guess we wait for #4 to be merged, before working on this further? Once we have PHPUnit 9, we get better type inference too \o/

**Review by Ocramius — APPROVED** [review]
LGTM :+1:

**Comment by Ocramius** [comment]
Excellent work, thank you, @diwms!

**Pull request merged**
